File: backend/src/models/Warehouse.js
import { pool } from "../libs/db.js";

export async function createWarehouse(ownerId, categories) {
  const connection = await pool.getConnection();
  try {
    await connection.beginTransaction();

    // 1. Create Warehouse
    const [result] = await connection.query(
      `INSERT INTO warehouse (categories, owner_id) VALUES (?, ?)`,
      [categories, ownerId]
    );
    const warehouseId = result.insertId;

    // 2. Update User
    await connection.query(
      `UPDATE users SET warehouse_id = ? WHERE user_id = ?`,
      [warehouseId, ownerId]
    );

    await connection.commit();
    return warehouseId;
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
}

export async function getWarehouseByToken(token) {
  const [rows] = await pool.query(
    `SELECT * FROM warehouse WHERE invitation_token = ? AND token_expires_at > NOW()`,
    [token]
  );
  return rows[0];
}

export async function getWarehouseById(warehouseId) {
  const [rows] = await pool.query(
    `SELECT * FROM warehouse WHERE warehouse_id = ?`,
    [warehouseId]
  );
  return rows[0];
}

export async function getWarehouseByOwnerId(ownerId) {
  const [rows] = await pool.query(
    `SELECT * FROM warehouse WHERE owner_id = ?`,
    [ownerId]
  );
  return rows[0];
}

export async function updateWarehouseToken(warehouseId, token, expiresAt) {
  await pool.query(
    `UPDATE warehouse SET invitation_token = ?, token_expires_at = ? WHERE warehouse_id = ?`,
    [token, expiresAt, warehouseId]
  );
}

File: services/api-gateway/src/domain/models/warehouse.py

from pydantic import BaseModel
from typing import Optional
from datetime import datetime

class Warehouse(BaseModel):
    warehouse_id: Optional[int] = None
    categories: Optional[str] = None
    owner_id: Optional[int] = None
    invitation_token: Optional[str] = None
    token_expires_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    
class InvitationToken(BaseModel):
    token: str
    warehouse_id: int
    created_by: int
    expires_at: datetime
    
    def is_expired(self):
        if self.expires_at < datetime.utcnow(): return True
        return False